### PR TITLE
#387 - `StateService` keeps state and nonce in-memory which prevents HA of oxd.

### PR DESCRIPTION
**Pull request by duttarnab**
#387 - `StateService` keeps state and nonce in-memory which prevents HA of oxd.
https://github.com/GluuFederation/oxd/issues/387

### CODE REVIEW

File: oxd-server/src/main/java/org/gluu/oxd/server/persistence/SqlPersistenceServiceImpl.java [hunk 1]
```diff
@@ -41,12 +41,16 @@ private void createSchema() {
...
             conn.setAutoCommit(false);
 
-            PreparedStatement createRpTable = conn.prepareStatement(
-                    "create table if not exists rp(id varchar(36) primary key, data varchar(65534))");
-            createRpTable.executeUpdate();
-            createRpTable.close();
+            Statement stmt = conn.createStatement();
 
+            stmt.addBatch("create table if not exists rp(id varchar(36) primary key, data varchar(65534))");
+            stmt.addBatch("create table if not exists expired_objects( key varchar(50), value varchar(50), type varchar(5), iat TIMESTAMP, exp TIMESTAMP)");
```
Comment (yuriyz) [inline oxd-server/src/main/java/org/gluu/oxd/server/persistence/SqlPersistenceServiceImpl.java]: value varchar(65534), type varchar(20)

File: oxd-common/src/main/java/org/gluu/oxd/common/ExpiredObject.java [hunk 2]
```diff
@@ -0,0 +1,63 @@
+package org.gluu.oxd.common;
+
+import java.sql.Timestamp;
```
[the remark this answers is not in the document]
Comment (duttarnab) [reply oxd-common/src/main/java/org/gluu/oxd/common/ExpiredObject.java]: Changed to long.

File: oxd-server/src/main/java/org/gluu/oxd/server/Utils.java [hunk 3]
```diff
@@ -108,4 +113,21 @@ public static String mapAsStringWithEncodedValues(Map<String, String> p_map) {
             return "";
         }
     }
+
+    public static ExpiredObject createExpiredObject(String key, ExpiredObjectType expiredObjectType, int expiredObjectExpirationInMins) {
```
[the remark this answers is not in the document]
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/Utils.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/Utils.java [hunk 4]
```diff
@@ -108,4 +113,21 @@ public static String mapAsStringWithEncodedValues(Map<String, String> p_map) {
...
+        if (!Strings.isNullOrEmpty(key)) {
```
[the remark this answers is not in the document]
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/Utils.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/persistence/PersistenceServiceImpl.java [hunk 5]
```diff
@@ -30,17 +38,41 @@ public void create() {
...
+
             return new SqlPersistenceServiceImpl(sqlProvider);
         } else if ("redis".equalsIgnoreCase(storage)) {
             return new RedisPersistenceService(configurationService.getConfiguration());
         }
         throw new RuntimeException("Failed to create persistence provider. Unrecognized storage specified: " + storage + ", full configuration: " + configurationService.get());
     }
 
+    public void setTimerForDBCleanUpTask() {
+        ScheduledExecutorService scheduledExecutorService = Executors.newScheduledThreadPool(1);
```
[the remark this answers is not in the document]
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/persistence/PersistenceServiceImpl.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/persistence/RedisPersistenceService.java [hunk 6]
```diff
@@ -55,6 +58,24 @@ public boolean create(Rp rp) {
...
+    public boolean createExpiredObject(ExpiredObject obj) {
+        try {
+            int objectExpirationInMinutes = 0;
+            if(obj.getType() == ExpiredObjectType.STATE){
+                objectExpirationInMinutes = configuration.getStateExpirationInMinutes();
+            } else if(obj.getType() == ExpiredObjectType.NONCE){
+                objectExpirationInMinutes = configuration.getNonceExpirationInMinutes();
+            }
+
+            put(objectExpirationInMinutes * 60, obj.getKey(), obj.getType().getValue());
```
[the remark this answers is not in the document]
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/persistence/RedisPersistenceService.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/persistence/RedisPersistenceService.java [hunk 7]
```diff
@@ -71,6 +92,23 @@ public Rp getRp(String oxdId) {
         return MigrationService.parseRp(get(oxdId));
     }
 
+    public ExpiredObject getExpiredObject(String key) {
+        String expiredObjectType = (String) redisProvider.get(key);
+        if(!Strings.isNullOrEmpty(expiredObjectType)){
+            ExpiredObject expiredObject = new ExpiredObject();
+            expiredObject.setKey(key);
+            expiredObject.setValue(key);
```
[the remark this answers is not in the document]
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/persistence/RedisPersistenceService.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/persistence/RedisPersistenceService.java [hunk 8]
```diff
@@ -71,6 +92,23 @@ public Rp getRp(String oxdId) {
         return MigrationService.parseRp(get(oxdId));
     }
 
+    public ExpiredObject getExpiredObject(String key) {
+        String expiredObjectType = (String) redisProvider.get(key);
```
[the remark this answers is not in the document]
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/persistence/RedisPersistenceService.java]: Done

File: oxd-server/src/main/java/org/gluu/oxd/server/persistence/RedisPersistenceService.java [hunk 9]
```diff
@@ -71,6 +92,23 @@ public Rp getRp(String oxdId) {
...
+        if(!Strings.isNullOrEmpty(expiredObjectType)){
```
[the remark this answers is not in the document]
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/persistence/RedisPersistenceService.java]: Storing  json of ExpiredObject in value attribute. So Type is now available. Tested locally and by running test case.

File: oxd-server/src/main/java/org/gluu/oxd/server/persistence/SqlPersistenceServiceImpl.java (hunk 1 above)
Comment (duttarnab) [reply oxd-server/src/main/java/org/gluu/oxd/server/persistence/SqlPersistenceServiceImpl.java]: Done

**Pull request merged**